Implement Shortlimitholdem

jake-bickle opened the pull request:
This branch implements Shortlimitholdem game and resolves #1 . 
I've implemented a few test cases to make sure that the few rule changes that differ from regular LimitHoldem are enforced correctly.

Running `python3 -m unittest discover` on `master` and `shortlimitholdem` branches shows that there are no test failures that didn't already exist on the `master` branch.

This branch also implements a bug fix within LimitHoldem, and by extension NoLimitHoldem, where the lowest rank (in these games, any 2 card) was not considered to be apart of a straight hand.

jake-bickle committed:
Add code-workspace to gitignore

jake-bickle committed:
Implement init_short_deck()

jake-bickle committed:
Create shortlimitholdem package. Implement dealer.py

jake-bickle committed:
Fix the lowest rank not being concidered in a straight in limitholdem/utils.py

jake-bickle committed:
Switched a few constants to be class attributes rather than instance attributes in Hand

jake-bickle committed:
Implement ShortHand class

jake-bickle committed:
Create test cases for ShortHand

jake-bickle committed:
Removed judger and player from shortlimitholdem, replaced with identical judger and player from limitholdem

jake-bickle committed:
Removed pdb in _get_straight_cards

jake-bickle committed:
Add DEALER_CLASS and JUDGER_CLASS to LimitholdemGame to make modifying the rules for subclasses easier

jake-bickle committed:
Implement ShortLimitHoldemGame

jake-bickle committed:
Add test_ace_is_high_card

jake-bickle committed:
Remove redundent copy/pasted code in ShortLimitHoldemDealer by making it a subclass of LimitHoldem's dealer

jneckar commented on rlcard/games/shortlimitholdem/utils.py:
```diff
@@ -0,0 +1,83 @@
+        lowest_card = cards[0]
+        if highest_card[1] == 'A':
+            cards.insert(0, highest_card)
+        elif lowest_card[1] == 'A':
```
Why is this necessary?

jneckar commented on rlcard/games/shortlimitholdem/utils.py:
```diff
@@ -0,0 +1,83 @@
+        lowest_card = cards[0]
+        if highest_card[1] == 'A':
+            cards.insert(0, highest_card)
+        elif lowest_card[1] == 'A':
+            cards.append(lowest_card)
```
I don't understand why this is necessary, ace should already be in highest card position

jake-bickle replied:
That code fixed an issue with the evaluator earlier, but I think I fixed the underlying bug that was causing other issues in the first place so it's indeed redundant. I'll fix it up.

jneckar commented on rlcard/games/shortlimitholdem/utils.py:
```diff
@@ -0,0 +1,83 @@
+    '''
+    RANK_TO_STRING = {6: "6", 7: "7", 8: "8", 9: "9", 10: "T",
+                      11: "J", 12: "Q", 13: "K", 14: "A"}
+    RANK_LOOKUP = "A6789TJQKA"
```
Is there a reason for the low A? It seems like the straight evaluator already checks for an ace and inserts it as a low card so I don't think it's necessary

jake-bickle replied:
The straight evaluator uses RANK_LOOKUP as apart of its logic. It takes a string of ranks and sees if it exists within RANK_LOOKUP. If the low A was not there, the string "A6789" would not exist and incorrectly return false.

jneckar commented on rlcard/games/shortlimitholdem/utils.py:
```diff
@@ -0,0 +1,83 @@
+            self.category = 6
+            #Full house
+            self.best_five = self._get_Fullhouse_cards()
+        elif self._has_three():
+            self.category = 5
+            #Three of a Kind
+            self.best_five = self._get_Three_of_a_kind_cards()
+        elif self._has_straight(self.all_cards):
+            self.category = 4
+            #Straight
```
Both rule variations exist, but in the more commonly played version, a straight beats 3 of a kind. Let's swap these hand rankings.

jneckar commented on tests/utils/test_shortlimitholdem_utils.py:
```diff
@@ -0,0 +1,37 @@
+from rlcard.games.shortlimitholdem.utils import ShortHand
+
+class TestShortLimitHoldemUtils(unittest.TestCase):
+    ''' ShortLimitHoldem is not much different from regular LimitHoldem. However, in LimitHoldem,
+    aces only count as a high card, while they count as a low and high card in Short.
+    Because of this, the majority of these tests make sure this works.
```
Aces should count as either high or low in any holdem variation, if that wasn't the case for limit holdem then it was a bug in their implementation

jake-bickle replied:
Interesting. I'll fix it up in Limitholdem which should then work for Short as well as No Limit.

jneckar reviewed (changes requested):
Great work, there were a few small things I noticed. Check comments in /shortlimitholdem/utils.py

jake-bickle committed:
Swap straight and three-of-a-kind rankings in evaluateHand

jake-bickle committed:
Fix aces not counting as high and low in straights in LimitHoldem

jake-bickle committed:
Remove _get_straight_cards implementation in ShortHand

jake-bickle committed:
Fix up tests in TestShortLimitHoldemUtils

jake-bickle commented:
I removed the redundancies and fixed the underlying bug of aces not counting as highs and lows in straights for LimitHoldem, which also extends to NoLimit and now Short as well. 
I couldn't find any pretrained models for these variants, so I don't think there'll be any side effects to fixing this bug.

jake-bickle committed:
Refactored ShortHand:evaluateHand() to be much simpler

jake-bickle committed:
Refactored ShortHand:evaluateHand() to be much simpler

jake-bickle committed:
Merge branch 'shortlimitholdem' of https://github.com/jake-bickle/rlcard into shortlimitholdem

jake-bickle committed:
Remove comment from TestShortLimitHoldemUtils
